src/Lib.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Lib (app) where
+
+import Control.Monad
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Except (ExceptT (..), runExceptT, withExceptT)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import Data.Coerce (coerce)
+import Data.Either (isLeft)
+import Data.Maybe (fromMaybe)
+import Database.SQLite.Simple
+import Database.SQLite.Simple.FromField (FromField (..))
+import Database.SQLite.Simple.ToField (ToField (..))
+import GHC.IO.Exception (ExitCode (ExitSuccess))
+import Network.Wai.Handler.Warp (run)
+import Servant
+import qualified System.Directory as Dir
+import System.FilePath ((</>))
+import qualified System.Process as Proc
+
+-- | Top-level API. Examples:
+--   localhost:8081/ps
+--   localhost:8081/procps/ps?sys=aarch64-linux
+--  TODO API documentation
+type API =
+  Get '[JSON] NoContent
+    :<|> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+    :<|> Capture "package-name" PackageName :> Capture "binary-name" BinaryName :> QueryParam "sys" System :> Get '[OctetStream] ByteString
+
+newtype BinaryName = BinaryName {unBinaryName :: String}
+  deriving newtype (ToField, FromHttpApiData)
+
+newtype PackageName = PackageName {unPackageName :: String}
+  deriving newtype (Eq, Ord, FromField, FromHttpApiData, ToField)
+
+-- TODO just make this a string?
+-- TODO isn't this an architecture more than a system? nix calls it a system I guess
+data System
+  = X86_64_Linux
+  | AARCH_64_Linux
+
+instance Show System where
+  show X86_64_Linux = "x86_64-linux"
+  show AARCH_64_Linux = "aarch64-linux"
+
+instance FromHttpApiData System where
+  parseQueryParam "aarch64" = pure AARCH_64_Linux
+  parseQueryParam "x86_64-linux" = pure X86_64_Linux
+  parseQueryParam sys = Left $ "Unknown system architecture: " <> sys
+
+-- | A Triplet uniquely identifies a binary in nixpkgs.
+-- A Triplet is not a proof that that binary exists and is buildable, though.
+-- TODO better name for this than triplet?
```

Comment:
How about `ProgramPath` ?
This sounds some deterministic but not completely.

Reply:
Hmm, maybe that's confusing because it could also mean `/nix/store/abcabc-whatever/bin/program`?

Reply:
Hm, maybe.
How about `ProgramPointer` then?

Reply:
The argument for Triplet is that, just like in mathematics, it's simply a point in a space. It may or may not correspond to a buildable binary, but we don't actually know that yet, and in fact, we need to store and cache Triplets regardless of whether they are buildable binaries/programs. I feel like when we say it `Points` or `Selects` or `Identifies` or any of the other names I can think of, we are making promises too early.

Of course, it's not just any space, it's the space of things that we can ask Nix to try building, and I don't know how to pithily express that. Maybe we should go with `ProgramSelector`.

Reply:
That is very suggestive. So the key here is, it is potentially "buildable".

In that aspect, this triplet is something being consumed when somebody tries to build something, which is potentially failing. It is "build" rather than "program". Following that idea, here are some ideas: `BuildPointer`, `BuildSelector`, `BuildDirection`, `BuildIndication`, `BuildInstruction`, `BuildRecipe`...

I am okay with `ProgramSelector` but if buildability matters as you say, naming after it makes much more sense.

Reply:
Unfortunately, "potentially buildable" is not quite the whole story; only the `Platform` x `Package` part affects buildability, but after building it also needs to produce the expected binary. I also thought about `BuildRecipe`, but I feel like that doesn't tell the whole story.

How do you feel about `BinaryTriplet`, would you be okay with `Triplet` if it's qualified like that? I feel like that nicely evokes that it's a triplet to identify a Binary, without implying that it's necessarily buildable.

Reply:
Yeah, I am okay with `BinaryTriplet` or `Triplet` if you feel more comfortable with it than the other candidates.

I guess the word "triplet" means a bit more to you than to me. (to me, that sounded a "dog")
Now it is getting to sound more binary or program.

Reply:
45ad7d57e7cf4fb74248a97a98a8629ad20aa059